### Notebook 1: Initial Exploration of Avocado Data

This notebook contains the first step of this Price Elasticity Project: reading in the dataset I will need, doing any cleaning and feature analysis, and exploring any initial correlations or relationships.

Import the needed libraries

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

Read in the avocado dataset

In [9]:
avodata = pd.read_csv('data/avocado.csv', index_col = 0 )
avodata.head(10)

Date  AveragePrice  Total Volume     4046       4225    4770  \
0  2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1  2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2  2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3  2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4  2015-11-29          1.28      51039.60   941.48   43838.39   75.78   
5  2015-11-22          1.26      55979.78  1184.27   48067.99   43.61   
6  2015-11-15          0.99      83453.76  1368.92   73672.72   93.26   
7  2015-11-08          0.98     109428.33   703.75  101815.36   80.00   
8  2015-11-01          1.02      99811.42  1022.15   87315.57   85.34   
9  2015-10-25          1.07      74338.76   842.40   64757.44  113.00   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49          0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14          0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76          0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69          0.0  conventional  2015  Albany  
5     6683.91     6556.47      127.44          0.0  conventional  2015  Albany  
6     8318.86     8196.81      122.05          0.0  conventional  2015  Albany  
7     6829.22     6266.85      562.37          0.0  conventional  2015  Albany  
8    11388.36    11104.53      283.83          0.0  conventional  2015  Albany  
9     8625.92     8061.47      564.45          0.0  conventional  2015  Albany

In [11]:
avodata.tail(10)

Date  AveragePrice  Total Volume     4046     4225    4770  \
2   2018-03-11          1.56      22128.42  2162.67  3194.25    8.93   
3   2018-03-04          1.54      17393.30  1832.24  1905.57    0.00   
4   2018-02-25          1.57      18421.24  1974.26  2482.65    0.00   
5   2018-02-18          1.56      17597.12  1892.05  1928.36    0.00   
6   2018-02-11          1.57      15986.17  1924.28  1368.32    0.00   
7   2018-02-04          1.63      17074.83  2046.96  1529.20    0.00   
8   2018-01-28          1.71      13888.04  1191.70  3431.50    0.00   
9   2018-01-21          1.87      13766.76  1191.92  2452.79  727.94   
10  2018-01-14          1.93      16205.22  1527.63  2981.04  727.01   
11  2018-01-07          1.62      17489.58  2894.77  2356.13  224.53   

    Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
2     16762.57    16510.32      252.25          0.0  organic  2018   
3     13655.49    13401.93      253.56          0.0  organic  2018   
4     13964.33    13698.27      266.06          0.0  organic  2018   
5     13776.71    13553.53      223.18          0.0  organic  2018   
6     12693.57    12437.35      256.22          0.0  organic  2018   
7     13498.67    13066.82      431.85          0.0  organic  2018   
8      9264.84     8940.04      324.80          0.0  organic  2018   
9      9394.11     9351.80       42.31          0.0  organic  2018   
10    10969.54    10919.54       50.00          0.0  organic  2018   
11    12014.15    11988.14       26.01          0.0  organic  2018   

              region  
2   WestTexNewMexico  
3   WestTexNewMexico  
4   WestTexNewMexico  
5   WestTexNewMexico  
6   WestTexNewMexico  
7   WestTexNewMexico  
8   WestTexNewMexico  
9   WestTexNewMexico  
10  WestTexNewMexico  
11  WestTexNewMexico

My first observation is that this is weekly price data. That is good to know. From the dataset description on Kaggle, the column descriptions are:
- Date: date of observation
- AveragePrice: the average price of a single avocado
- Total Volume: the total number of avocados sold
- 4046: the total number of avocados with PLU 4046 sold
- 4225: the total number of avocados with PLU 4225 sold
- 4770: the total number of avocados with PLU 4770 sold
- Total Bags: Number of bags the avocados were sold in
- Small Bags: Number of small bags of avocados delivered
- Large Bags: Number of large bags of avocados delivered
- XL Bags: Number of extra large bags of avocados delivered
- Type: type of avocado, either organic or conventional
- year: the year of the observation
- region: the region where the avocados were sold

In [14]:
# Let's take a look at the different regions available to us:
avodata['region'].unique()

array(['Albany', 'Atlanta', 'BaltimoreWashington', 'Boise', 'Boston',
       'BuffaloRochester', 'California', 'Charlotte', 'Chicago',
       'CincinnatiDayton', 'Columbus', 'DallasFtWorth', 'Denver',
       'Detroit', 'GrandRapids', 'GreatLakes', 'HarrisburgScranton',
       'HartfordSpringfield', 'Houston', 'Indianapolis', 'Jacksonville',
       'LasVegas', 'LosAngeles', 'Louisville', 'MiamiFtLauderdale',
       'Midsouth', 'Nashville', 'NewOrleansMobile', 'NewYork',
       'Northeast', 'NorthernNewEngland', 'Orlando', 'Philadelphia',
       'PhoenixTucson', 'Pittsburgh', 'Plains', 'Portland',
       'RaleighGreensboro', 'RichmondNorfolk', 'Roanoke', 'Sacramento',
       'SanDiego', 'SanFrancisco', 'Seattle', 'SouthCarolina',
       'SouthCentral', 'Southeast', 'Spokane', 'StLouis', 'Syracuse',
       'Tampa', 'TotalUS', 'West', 'WestTexNewMexico'], dtype=object)

In [26]:
# Which regions bought the most avocados?
avodata.groupby('region').sum().sort_values('total_volume', ascending= False)[:10]

averageprice  total_volume          4046          4225  \
region                                                                 
TotalUS             445.83  5.864740e+09  2.054936e+09  2.015012e+09   
West                430.01  1.086779e+09  3.985905e+08  3.007589e+08   
California          471.62  1.028982e+09  3.989670e+08  3.514914e+08   
SouthCentral        372.22  1.011280e+09  5.350414e+08  2.204500e+08   
Northeast           541.45  7.132809e+08  3.499121e+07  4.744847e+08   
Southeast           472.53  6.152384e+08  3.396871e+08  9.086044e+07   
GreatLakes          452.43  5.896425e+08  9.371701e+07  2.672577e+08   
Midsouth            474.81  5.083494e+08  1.104306e+08  2.223731e+08   
LosAngeles          411.01  5.078965e+08  2.238794e+08  1.021340e+08   
Plains              485.54  3.111885e+08  1.435114e+08  8.733534e+07   

                      4770    total_bags    small_bags    large bags  \
region                                                                 
TotalUS       1.561752e+08  1.638601e+09  1.243561e+09  3.735843e+08   
West          2.035607e+07  3.670736e+08  2.227093e+08  1.435317e+08   
California    3.184105e+07  2.466822e+08  2.290977e+08  1.231194e+07   
SouthCentral  2.239579e+07  2.333924e+08  1.848155e+08  4.593684e+07   
Northeast     6.816645e+06  1.969884e+08  1.736230e+08  2.205605e+07   
Southeast     3.163231e+06  1.815276e+08  1.069033e+08  7.210761e+07   
GreatLakes    5.007597e+07  1.785756e+08  1.232875e+08  4.889564e+07   
Midsouth      1.903196e+07  1.565136e+08  1.343644e+08  2.077087e+07   
LosAngeles    1.361869e+07  1.682645e+08  1.557261e+08  8.463145e+06   
Plains        2.494474e+06  7.784720e+07  6.876062e+07  7.973644e+06   

                  xl_bags    year  
region                             
TotalUS       21455230.54  681458  
West            832634.65  681458  
California     5272577.91  681458  
SouthCentral   2640111.14  681458  
Northeast      1309367.56  681458  
Southeast      2516762.93  681458  
GreatLakes     6392523.96  681458  
Midsouth       1378317.13  681458  
LosAngeles     4075279.95  681458  
Plains         1112937.29  681458

In [16]:
# I'm going to rename the columns so there are no spaces and everything is lowercase:
avodata.columns = avodata.columns.str.lower()

avodata.rename(columns={
    'total volume': 'total_volume',
    'total bags'  : 'total_bags',
    'small bags'  : 'small_bags',
    'xlarge bags' : 'xl_bags',
}, inplace=True)
avodata.head()

date  averageprice  total_volume     4046       4225    4770  \
0  2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1  2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2  2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3  2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4  2015-11-29          1.28      51039.60   941.48   43838.39   75.78   

   total_bags  small_bags  large bags  xl_bags          type  year  region  
0     8696.87     8603.62       93.25      0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49      0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14      0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76      0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69      0.0  conventional  2015  Albany

In [27]:
# Let me see if there are any missing values
avodata.isna().sum()

date            0
averageprice    0
total_volume    0
4046            0
4225            0
4770            0
total_bags      0
small_bags      0
large bags      0
xl_bags         0
type            0
year            0
region          0
dtype: int64

In [ ]:
# Looks like a clean dataset

I'd like to look at some initial correlations with price elasticity before I start modeling. In order to do that, I first need to calculate price elasticity by region over the course of each week and add that to the dataset.

In [29]:
avodata.groupby('region')